module/scmi_reset/src/mod_scmi_reset.c

```diff
@@ -0,0 +1,447 @@
+ */
+static bool service_is_permitted(const struct mod_scmi_reset_device *device,
+                                enum mod_scmi_reset_permissions request)
+{
+    return device->permissions & request;
+}
+
+static int reset_attributes_handler(fwk_id_t service_id,
+                                    const uint32_t *payload)
+{
+    const struct mod_scmi_reset_device *reset_device = NULL;
+    struct scmi_reset_attributes_a2p params = { 0 };
+    struct scmi_reset_attributes_p2a outmsg = {
+        .status = SCMI_GENERIC_ERROR,
+    };
+    size_t outmsg_size = sizeof(outmsg.status);
+    int status = FWK_SUCCESS;
+
+    params = *(const struct scmi_reset_attributes_a2p *)payload;
+
+    status = get_reset_device(service_id, params.domain_id, &reset_device);
+    if (status != FWK_SUCCESS) {
+        outmsg.status = SCMI_NOT_FOUND;
+        goto exit;
+    }
+
+    if (!service_is_permitted(reset_device, MOD_SCMI_RESET_PERM_ATTRIBUTES)) {
+        outmsg.status = SCMI_DENIED;
+        goto exit;
+    }
+
+    /*
+     * TODO: get Async/Sync support info from module reset
+     * Currently: no support for async reset, no support for notification
+     */
+    outmsg.flags &= ~SCMI_RESET_DOMAIN_ATTR_ASYNC;
+    outmsg.flags &= ~SCMI_RESET_DOMAIN_ATTR_NOTIF;
+
+    /* TODO: Get reset latency from reset driver */
+    outmsg.latency = SCMI_RESET_DOMAIN_ATTR_UNK_LAT;
+
+    strncpy((char *)outmsg.name, fwk_module_get_name(reset_device->element_id),
+            sizeof(outmsg.name) - 1);
+
+    outmsg.status = SCMI_SUCCESS;
+    outmsg_size = sizeof(outmsg);
+
+exit:
+    scmi_reset_ctx.scmi_api->respond(service_id, &outmsg, outmsg_size);
+
+    return status;
+}
+
+static int reset_request_handler(fwk_id_t service_id,
+                                 const uint32_t *payload)
+{
+    int status = FWK_SUCCESS;
+    struct scmi_reset_request_a2p params = { 0 };
+    struct scmi_reset_request_p2a outmsg = {
+        .status = SCMI_NOT_FOUND
+    };
+    size_t outmsg_size = sizeof(outmsg.status);
+    bool async_reset = false;
+    bool assert_reset = false;
+    bool deassert_reset = false;
+    const struct mod_reset_drv_api *reset_api = NULL;
+    const struct mod_scmi_reset_device *reset_device = NULL;
+
+    params = *(const struct scmi_reset_request_a2p *)payload;
+
+    if (params.domain_id >= SCMI_RESET_DOMAIN_COUNT_MASK)
+        goto exit;
+
+    status = get_reset_device(service_id, params.domain_id, &reset_device);
+    if (status != FWK_SUCCESS) {
+        goto exit;
+    }
+
+    if (!service_is_permitted(reset_device, MOD_SCMI_RESET_PERM_RESET)) {
+        outmsg.status = SCMI_DENIED;		/* Should be obfuscated? */
+        goto exit;
+    }
+
+    async_reset = params.flags & SCMI_RESET_DOMAIN_ASYNC;
+    assert_reset = params.flags &
+	           (SCMI_RESET_DOMAIN_ASSERT | SCMI_RESET_DOMAIN_AUTO);
+    deassert_reset = (params.flags & SCMI_RESET_DOMAIN_AUTO) ||
+                     !(params.flags & SCMI_RESET_DOMAIN_ASSERT);
+
```

Comment:
The way the protocol is setup is:
Assert -> Only assert the line. No choice of state (state ignored), since a line can only be asserted or deasserted.
Deassert -> Only deassert the line.
Auto -> This is a specific mode for devices which can handle auto reset. Such devices will typically have a set of bit having different reset states where writing to the bit will reset the device and then bring it out of reset. It might even do some device specific behavior.

Hence the SCMI protocol should just transparently pass either:
1) Assert with no state
2) Deassert with no state
3) Auto with state to the reset domain driver.
SCMI lets the reset domain driver decide how to implement the reset for the domain, since SCMI is a common protocol serving a wide variety of platforms and so cannot take platform specific decisions.

Reply:
ok, thanks. I understand now.

Reply:
Comment should be addressed, now.